python/triton/language/semantic.py

```diff
@@ -795,96 +795,196 @@ def str_to_eviction_policy(eviction_policy):
 
-    if not mask:
-        if other:
+        # Check `mask`, `other`, `boundary_check`, and `padding` arguments
+        if not mask and other:
             raise ValueError("`other` cannot be provided without `mask`")
-        return tl.tensor(builder.create_load(ptr.handle, cache, eviction, is_volatile),
-                         dst_ty)
-    else:
-        return tl.tensor(builder.create_masked_load(ptr.handle,
-                                                    mask.handle,
-                                                    other.handle if other else None,
-                                                    cache, eviction, is_volatile),
-                         dst_ty)
+        if padding_option or boundary_check:
+            raise ValueError("`padding_option` or `boundary_check` argument is not supported for loading a tensor of"
+                             "pointers or loading a scalar. Because the compiler does not know the boundary; please "
+                             "use tiled mode load instead")
+
+        # For a pointer of scalar, check the type of `mask` and `other`
+        if not ptr.type.is_block():
+            if mask and mask.type.is_block():
+                raise ValueError("Mask argument cannot be block type if pointer argument is not a block")
+            if other and other.type.is_block():
+                raise ValueError("Other argument cannot be block type if pointer argument is not a block")
+
+        # Make `mask` and `other` into the same shape as `ptr`
+        if ptr.type.is_block():
+            if mask:
+                mask = broadcast_impl_shape(mask, ptr.type.get_block_shapes(), builder)
+            if other:
+                other = broadcast_impl_shape(other, ptr.type.get_block_shapes(), builder)
+
+        # Get `pointer_type<elt_ty>` and `elt_ty`
+        ptr_ty = ptr.type.scalar
+        elt_ty = ptr_ty.element_ty
+
+        # Treat `pointer_type<tl.int1>` as `pointer_type<tl.int8>`
+        if elt_ty == tl.int1:
+            elt_ty = tl.int8
+            ptr_ty = tl.pointer_type(elt_ty, ptr_ty.address_space)
+            ptr = cast(ptr, ptr_ty, builder)
+
+        # Cast `other` into `ele_ty` type
+        if other:
+            other = cast(other, elt_ty, builder)
+
+        # Create loaded result type `dst_ty`
+        if ptr.type.is_block():
+            shape = ptr.type.get_block_shapes()
+            dst_ty = tl.block_type(elt_ty, shape)
+        else:
+            # Load by de-referencing the pointer of scalar
+            dst_ty = elt_ty
+
+        # Build IR for non-tiled load
+        if not mask:
+            return tl.tensor(builder.create_load(ptr.handle, cache, eviction, is_volatile), dst_ty)
+        else:
+            return tl.tensor(builder.create_masked_load(ptr.handle, mask.handle, other.handle if other else None,
+                                                        cache, eviction, is_volatile), dst_ty)
 
 
 def store(ptr: tl.tensor,
```

Comment:
ditto; we want to pick between some _unstructured_store and _structured_store depending on the type of pointer. The check that value.type == pointer.pointee_type.type can be mutualized.

Reply:
Fixed. I separate `load` into `_load_block_pointer` and `_load_legacy`. Same for `store.`